lib/spack/spack/database.py

```diff
@@ -299,12 +299,9 @@ def __reduce__(self):
                 database.  If it is a spec, we'll evaluate
                 ``spec.satisfies(query_spec)``
 
-            known (bool or None): Specs that are "known" are those
-                for which Spack can locate a ``package.py`` file -- i.e.,
-                Spack "knows" how to install them.  Specs that are unknown may
-                represent packages that existed in a previous version of
-                Spack, but have since either changed their name or
-                been removed
```

Comment:
> Can we first do a fix that doesn't change the signature?

I took a look at that. It doesn't make much sense to me to do it.  Rationale:
- The bug is essentially an oversight (a missing commit probably). The idea was to turn `known=` into the state of this PR, but in the codebase it is a weird half-state where it is treated as a callable and documented as a bool
- The bug was introduced before v0.21 so the signature here, when called as documented, will raise on v0.21 and v0.22 (the two versions we currently support). 
- This means there can be no script using `known=True`, or `known=False`, that is currently working as intended.
- Other parts of `Database.query` are not changed, so if `known=` is not used the script should continue working.

I think the fix that can be backported is essentially the behavior of `spack find -u`. Using `db.query(known=True)` was not possible before.

Feel free to remove the `v0.21` and `v0.22` labels if you don't think this should be backported.

Reply:
OK, i'm fine with a semi-breaking change.

My worry was if people for whatever reason copied the default arguments of `db.query(...)` to the call site, which would fail after this PR.